src/Sentry.Tunnel/SentryTunnelMiddleware.cs

```diff
@@ -0,0 +1,85 @@
+                context.Response.Headers.Add("Access-Control-Allow-Methods", new[] {"POST, OPTIONS"});
+                context.Response.Headers.Add("Access-Control-Allow-Credentials", new[] {"true"});
+                context.Response.StatusCode = 200;
+                return;
+            }
+
+            var httpClientFactory = context.RequestServices.GetRequiredService<IHttpClientFactory>();
+            var client = httpClientFactory.CreateClient("SentryTunnel");
+            client.DefaultRequestHeaders.UserAgent.Add(new ProductInfoHeaderValue("Sentry.NET Tunnel", GetType().Assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion));
+            var ms = new MemoryStream();
+            await context.Request.Body.CopyToAsync(ms);
+            ms.Position = 0;
+            using (var reader = new StreamReader(ms))
+            {
+                var header = await reader.ReadLineAsync();
+                if (string.IsNullOrWhiteSpace(header))
+                {
+                    context.Response.StatusCode = StatusCodes.Status400BadRequest;
+                    return;
+                }
+
+                try
+                {
+                    var headerJson = JsonSerializer.Deserialize<Dictionary<string, object>>(header);
+                    if (headerJson == null)
+                    {
+                        context.Response.StatusCode = StatusCodes.Status400BadRequest;
+                        await context.Response.WriteAsync("Invalid DSN JSON supplied");
+                        return;
+                    }
+                    if (headerJson.TryGetValue("dsn", out var dsnString) && Uri.TryCreate(dsnString.ToString(), UriKind.Absolute, out var dsn) && _allowedHosts.Contains(dsn.Host))
+                    {
+                        var projectId = dsn.AbsolutePath.Trim('/');
+                        ms.Position = 0;
+                        var responseMessage = await client.PostAsync($"https://{dsn.Host}/api/{projectId}/envelope/",
+                            new StreamContent(ms));
+                        context.Response.Headers["content-type"] = "application/json";
+                        context.Response.StatusCode = StatusCodes.Status200OK;
+                        await responseMessage.Content.CopyToAsync(context.Response.Body);
+                    }
+                }
+                catch(JsonException)
+                {
+                    context.Response.StatusCode = StatusCodes.Status400BadRequest;
+                    await context.Response.WriteAsync("Invalid DSN JSON supplied");
+                }
+                catch(ArgumentNullException)
+                {
+                    context.Response.StatusCode = StatusCodes.Status400BadRequest;
+                    await context.Response.WriteAsync("Received empty body");
+                }
```

Comment:
Do we need a content type `text/plain` or something when writing text out?

```suggestion
                catch (JsonException)
                {
                    context.Response.StatusCode = StatusCodes.Status400BadRequest;
                    await context.Response.WriteAsync("Invalid DSN JSON supplied");
                }
                catch (ArgumentNullException)
                {
                    context.Response.StatusCode = StatusCodes.Status400BadRequest;
                    await context.Response.WriteAsync("Received empty body");
                }
```